share/spack/spack-completion.bash

```diff
@@ -91,262 +101,325 @@ function _bash_completion_spack {
     fi
 }
 
+# Helper functions for subcommands
+
+_subcommands () {
+    spack commands
+}
+
+_all_packages () {
+    spack list
+}
+
+_all_resource_hashes () {
+    spack resource list --only-hashes
+}
+
+_installed_packages () {
+    spack --color=never find --no-groups
+}
+
+_installed_compilers () {
+    spack compilers | egrep -v "^(-|=)"
+}
+
+_providers () {
+    spack providers
+}
+
+_mirrors () {
+    spack mirror list | awk '{print $1}'
+}
+
+_repos () {
+    spack repo list | awk '{print $1}'
+}
+
+_tests () {
+    spack test -l
+}
+
+_environments () {
+    spack env list
+}
+
+_keys () {
+    spack gpg list
+}
+
+# Testing functions
+
+_test_vars () {
+    echo "-----------------------------------------------------"             >> temp
+    echo "Full line:                '$COMP_LINE'"                            >> temp
+    echo                                                                     >> temp
+    echo "Word list w/ flags:       $(_pretty_print COMP_WORDS[@])"          >> temp
+    echo "# words w/ flags:         '${#COMP_WORDS[@]}'"                     >> temp
+    echo "Cursor index w/ flags:    '$COMP_CWORD'"                           >> temp
+    echo                                                                     >> temp
+    echo "Word list w/out flags:    $(_pretty_print COMP_WORDS_NO_FLAGS[@])" >> temp
+    echo "# words w/out flags:      '${#COMP_WORDS_NO_FLAGS[@]}'"            >> temp
+    echo "Cursor index w/out flags: '$COMP_CWORD_NO_FLAGS'"                  >> temp
+    echo                                                                     >> temp
+    echo "Subfunction:              '$subfunction'"                          >> temp
+    if $list_options
+    then
+        echo "List options:             'True'"  >> temp
+    else
+        echo "List options:             'False'" >> temp
+    fi
+    echo "Current word:             '$cur'"  >> temp
+    echo "Previous word:            '$prev'" >> temp
+}
+
+# Pretty-prints one or more arrays
+# Syntax: _pretty_print array1[@] ...
+_pretty_print () {
+    for arg in $@
+    do
+        local array=("${!arg}")
+        echo -n "$arg: ["
+        printf   "'%s'" "${array[0]}"
+        printf ", '%s'" "${array[@]:1}"
+        echo "]"
+    done
+}
+
+complete -o bashdefault -o default -F _bash_completion_spack spack
```

Comment:
Git's tab completion also uses `-o bashdefault` which we weren't using before. I tested this with the system Bash on macOS (3.2.57) and Spack-installed Bash (5.0.11).